File: client/src/lib/api/projects.js
import { axios } from "./index"; // Import the configured axios

export const projectsApi = {
  // Create project
  createProject: async (data) => {
    const response = await axios.post("/projects", data);
    return response.data;
  },

  // List projects
  getProjects: async (params = {}) => {
    const response = await axios.get("/projects", { params });
    return response.data;
  },

  // Get single project
  getProject: async (projectId) => {
    const response = await axios.get(`/projects/${projectId}`);
    return response.data;
  },

  // Update project
  updateProject: async (projectId, data) => {
    const response = await axios.put(`/projects/${projectId}`, data);
    return response.data;
  },

  // Delete project
  deleteProject: async (projectId) => {
    await axios.delete(`/projects/${projectId}`);
  },

  // Get project progress
  getProjectProgress: async (projectId) => {
    const response = await axios.get(`/projects/${projectId}/progress`);
    return response.data;
  },

  // Get project members
  getMembers: async (projectId, params = {}) => {
    const response = await axios.get(`/projects/${projectId}/members`, {
      params,
    });
    return response.data;
  },

  // Add member to project
  addMember: async (projectId, userId) => {
    await axios.post(`/projects/${projectId}/members/${userId}`);
  },

  // Remove member from project
  removeMember: async (projectId, userId) => {
    await axios.delete(`/projects/${projectId}/members/${userId}`);
  },
};

File: server/app/api/v1/projects.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from uuid import UUID
from typing import List, Optional

from app.database import get_db
from app.models.user import User
from app.models.project import Project
from app.schemas.project import ProjectCreate, ProjectUpdate, ProjectResponse
from app.services.project_service import ProjectService
from app.dependencies import get_current_user, get_admin_user

router = APIRouter()


@router.post("/", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
async def create_project(
    project_data: ProjectCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_admin_user)
):
    """Create a new project (Admin only)."""
    project = await ProjectService.create_project(db, project_data, current_user)
    return project


@router.get("/", response_model=List[ProjectResponse])
async def list_projects(
    skip: int = 0,
    limit: int = 100,
    status_filter: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List projects: Users see only their projects, Admins see all."""
    projects = await ProjectService.get_projects(
        db,
        current_user=current_user,
        skip=skip,
        limit=limit,
        status=status_filter
    )
    return projects


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get project details: Users can only access their projects, Admins access all."""
    project = await ProjectService.get_project(db, project_id, current_user)
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found or access denied"
        )
    return project


@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: UUID,
    project_data: ProjectUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update project: Admins can update any, members can update projects they belong to."""
    project = await ProjectService.update_project(db, project_id, project_data, current_user)
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found"
        )
    return project


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_admin_user)
):
    """Delete project (Admin only)."""
    success = await ProjectService.delete_project(db, project_id, current_user)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found"
        )
    return None


@router.get("/{project_id}/progress")
async def get_project_progress(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get project progress: Users can only access their projects, Admins access all."""
    progress = await ProjectService.get_project_progress(db, project_id, current_user)
    if not progress:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found or access denied"
        )
    return progress